In [2]:
import pandas as pd

# EXERCISES

2. Given the following confusion matrix, evaluate (by hand) the model's performance.

|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |

   - In the context of this problem, what is a false positive?
   - In the context of this problem, what is a false negative?
   - How would you describe this model?


Exercise 2: 
    cat = negative class
    dog = positive class
    
    - False Positive = 7
    - False Negative = 13
    - Confusing. But I'm sure you mean this:
                TP = predicted dog & actual dog
                FP = predicted dog & actual cat
                TN = predicted cat & actual cat
                FN = predicted cat & actual dog 
    

Exercise 3: You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

Use the predictions dataset and pandas to help answer the following questions:

   3.1: An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?
   
   
   3.2: Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?


In [6]:
duckduckgoose= pd.read_csv('c3.csv')

In [7]:
duckduckgoose.head()

actual     model1  model2     model3
0  No Defect  No Defect  Defect  No Defect
1  No Defect  No Defect  Defect     Defect
2  No Defect  No Defect  Defect  No Defect
3  No Defect     Defect  Defect     Defect
4  No Defect  No Defect  Defect  No Defect

In [9]:
duckduckgoose.actual.value_counts()

No Defect    184
Defect        16
Name: actual, dtype: int64

Exercise 3.1.a: Which evaluation metric would be appropriate here?


Recall:

- allows us to minimize false negative & prioritize real positives
- want to flag defects
- recall = tp/ (tp+fn)
- defect = positive class, no defect = negative class

In [12]:
goosing_around= duckduckgoose[duckduckgoose.actual == 'Defect']
goosing_around.count()

actual    16
model1    16
model2    16
model3    16
dtype: int64

Exercise 3.1.b: Which model would be the best fit for this use case?

In [18]:
m1recall= (goosing_around.actual == goosing_around.model1).mean()
print(f' Model 1 Recall:{m1recall:.2%}')

 Model 1 Recall:50.00%


In [19]:
m2recall= (goosing_around.actual == goosing_around.model2).mean()
print(f' Model 2 Recall:{m2recall:.2%}')

 Model 2 Recall:56.25%


In [20]:
m3recall= (goosing_around.actual == goosing_around.model3).mean()
print(f' Model 3 Recall:{m3recall:.2%}')

 Model 3 Recall:81.25%


Answer: Model 3 has the highest recall and would be best fit for this case

Exercise 3.2.a: Which evaluation metric would be appropriate here? 


Answer: 
   - Priority is minimizing false positives, which implies wanting to prioritize high precision
   - ** Precision ** is the best metric for this task

Exercise 3.2.b: Which model would be the best fit for this use case?

In [21]:
duckin_donuts = duckduckgoose[duckduckgoose.model1 == 'Defect']

duckin_donuts

actual  model1     model2     model3
3    No Defect  Defect     Defect     Defect
30      Defect  Defect  No Defect     Defect
62   No Defect  Defect  No Defect  No Defect
65      Defect  Defect     Defect     Defect
70      Defect  Defect     Defect     Defect
135     Defect  Defect  No Defect     Defect
147     Defect  Defect  No Defect     Defect
163     Defect  Defect     Defect     Defect
194     Defect  Defect  No Defect     Defect
196     Defect  Defect  No Defect  No Defect

In [32]:
# Only need to identify model 1 defects
duckin_donuts = duckduckgoose[duckduckgoose.model1 == 'Defect']

m1precision = (duckin_donuts.actual == duckin_donuts.model1).mean()

print (f'Model 1 Presicion is {m1precision:.2%}')

Model 1 Presicion is 80.00%


In [33]:

# Only need to identify model 2 defects
duckin_donuts = duckduckgoose[duckduckgoose.model2 == 'Defect']

m2precision = (duckin_donuts.actual == duckin_donuts.model2).mean()

print (f'Model 2 Presicion is {m2precision:.2%}')

Model 2 Presicion is 10.00%


In [34]:
# Only need to identify model 3 defects
duckin_donuts = duckduckgoose[duckduckgoose.model3 == 'Defect']

m3precision = (duckin_donuts.actual == duckin_donuts.model3).mean()

print (f'Model 3 Presicion is {m3precision:.2%}')

Model 3 Presicion is 13.13%


Answer: Model 1 has the highest rate of precision and would be the best fit for this task

### Exercise 4:

You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

Several models have already been developed with the data, and you can find their results here.

Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

   1. In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?
   2. Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recommend?
   3. Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recommend?



In [36]:
paws = pd.read_csv('gives_you_paws.csv')
paws.head()

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

In [37]:
paws.actual.value_counts()

dog    3254
cat    1746
Name: actual, dtype: int64

Exercise 4.1: In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?

In [48]:
# Baseline prediction is usually set to find the most common occurance

# Baseline prediction = dog

# Baseline accuracy is percentage value of baseline prediction/ total

total = 3254+1746
baseline_acc = 3254/total
paws.baseline = paws.actual.value_counts().idxmax()
print(f'Baseline accuracy is {baseline_acc:.2%}')

Baseline accuracy is 65.08%


In [52]:
# Now, I need to find the model accuracy for all of the models to compare to the 
# baseline acc

m1acc= (paws.model1 == paws.actual).mean()
m2acc= (paws.model2 == paws.actual).mean()
m3acc= (paws.model3 == paws.actual).mean()
m4acc= (paws.model4 == paws.actual).mean()



print(f'Model 1 accuracy is {m1acc:.2%}')
print(f'Model 2 accuracy is {m2acc:.2%}')
print(f'Model 3 accuracy is {m3acc:.2%}')
print(f'Model 4 accuracy is {m4acc:.2%}')

Model 1 accuracy is 80.74%
Model 2 accuracy is 63.04%
Model 3 accuracy is 50.96%
Model 4 accuracy is 74.26%


Answer: Models 1 & 4 has a higher accuracy than the baseline

Exercise 4.2: Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recommend?

In [ ]:
#I should use recall to find the highest number of real positive cases
dog = positive class
cats = negative class

In [55]:
fido = paws[paws.actual == 'dog']
fido.head()

actual model1 model2 model3 model4
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
5    dog    dog    dog    dog    dog
8    dog    dog    cat    dog    dog

In [57]:
m1recall = (fido.actual == fido.model1).mean()
m2recall = (fido.actual == fido.model2).mean()
m3recall = (fido.actual == fido.model3).mean()
m4recall = (fido.actual == fido.model4).mean()

print(f'Model 1 Recall Accuracy:{m1recall:.2%}')
print(f'Model 2 Recall Accuracy:{m2recall:.2%}')
print(f'Model 3 Recall Accuracy:{m3recall:.2%}')
print(f'Model 4 Recall Accuracy:{m4recall:.2%}')

Model 1 Recall Accuracy:80.33%
Model 2 Recall Accuracy:49.08%
Model 3 Recall Accuracy:50.86%
Model 4 Recall Accuracy:95.57%


Answer: Model 4 has a recall of 95.57% and would be the best model for this task

Exercise 4.3: Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recommend?

In [62]:
# Since cats is my negative prediction, I need to switch from recall to precision

gatito = paws[paws.actual == 'cat']
gatito.head()

actual model1 model2 model3 model4
0     cat    cat    dog    cat    dog
4     cat    cat    cat    dog    dog
6     cat    cat    cat    cat    dog
7     cat    dog    cat    cat    dog
11    cat    cat    dog    cat    cat

In [59]:
baseline_acc = 1746/total
paws.baseline = paws.actual.value_counts().idxmax()
print(f'Baseline accuracy is {baseline_acc:.2%}')

Baseline accuracy is 34.92%


In [63]:
m1recall = (gatito.actual == gatito.model1).mean()
m2recall = (gatito.actual == gatito.model2).mean()
m3recall = (gatito.actual == gatito.model3).mean()
m4recall = (gatito.actual == gatito.model4).mean()

print(f'Model 1 Precision :{m1recall:.2%}')
print(f'Model 2 Precision :{m2recall:.2%}')
print(f'Model 3 Precision :{m3recall:.2%}')
print(f'Model 4 Precision :{m4recall:.2%}')

Model 1 Precision :81.50%
Model 2 Precision :89.06%
Model 3 Precision :51.15%
Model 4 Precision :34.54%


Answer: Model 2 has the best precision and would be the best fit for this task

### Exercise 5:

Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.

    sklearn.metrics.accuracy_score
    sklearn.metrics.precision_score
    sklearn.metrics.recall_score
    sklearn.metrics.classification_report



In [64]:
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score

#### Classification Reports:

In [67]:
pd.DataFrame(classification_report(paws.actual, paws.model1, 
                                   labels = ['cat', 'dog'], output_dict=True)).T

precision    recall  f1-score    support
cat            0.689772  0.815006  0.747178  1746.0000
dog            0.890024  0.803319  0.844452  3254.0000
accuracy       0.807400  0.807400  0.807400     0.8074
macro avg      0.789898  0.809162  0.795815  5000.0000
weighted avg   0.820096  0.807400  0.810484  5000.0000

In [68]:
pd.DataFrame(classification_report(paws.actual, paws.model2, 
                                   labels = ['cat', 'dog'], output_dict=True)).T

precision    recall  f1-score    support
cat            0.484122  0.890607  0.627269  1746.0000
dog            0.893177  0.490781  0.633479  3254.0000
accuracy       0.630400  0.630400  0.630400     0.6304
macro avg      0.688649  0.690694  0.630374  5000.0000
weighted avg   0.750335  0.630400  0.631310  5000.0000

In [69]:
pd.DataFrame(classification_report(paws.actual, paws.model3, 
                                   labels = ['cat', 'dog'], output_dict=True)).T

precision    recall  f1-score    support
cat            0.358347  0.511455  0.421425  1746.0000
dog            0.659888  0.508605  0.574453  3254.0000
accuracy       0.509600  0.509600  0.509600     0.5096
macro avg      0.509118  0.510030  0.497939  5000.0000
weighted avg   0.554590  0.509600  0.521016  5000.0000

In [70]:
pd.DataFrame(classification_report(paws.actual, paws.model4, 
                                   labels = ['cat', 'dog'], output_dict=True)).T

precision    recall  f1-score    support
cat            0.807229  0.345361  0.483755  1746.0000
dog            0.731249  0.955747  0.828560  3254.0000
accuracy       0.742600  0.742600  0.742600     0.7426
macro avg      0.769239  0.650554  0.656157  5000.0000
weighted avg   0.757781  0.742600  0.708154  5000.0000

In [79]:
print(f'Model 1 Accuracy Score:{accuracy_score(paws.actual, paws.model1)}')
print(f'Model 2 Accuracy Score:{accuracy_score(paws.actual, paws.model2)}')
print(f'Model 3 Accuracy Score:{accuracy_score(paws.actual, paws.model3)}')
print(f'Model 4 Accuracy Score:{accuracy_score(paws.actual, paws.model4)}')

Model 1 Accuracy Score:0.8074
Model 2 Accuracy Score:0.6304
Model 3 Accuracy Score:0.5096
Model 4 Accuracy Score:0.7426
